### PR TITLE
NEW Use a single-source-of-truth to determine version mapping

### PR DESCRIPTION
**Pull request by GuySartorelli**
## Issue
- #104

### CODE REVIEW

File: src/Console/CheckoutCommand.php [hunk 1]
```diff
@@ -11,50 +14,203 @@
...
+     * Builds a map of which Silverstripe CMS major release matches which major release of a module.
+     */
+    private function buildVersionMap(array $config, OutputInterface $output)
+    {
+        $output->writeln('Making new version map');
+        $map = [];
+        $versions = array_keys($config['versions']);
+
+        foreach ($versions as $version) {
+            $modulesUrl = "https://raw.githubusercontent.com/silverstripe/supported-modules/$version/modules.json";
```
Comment (GuySartorelli) [inline src/Console/CheckoutCommand.php:163]: This requires the supported-modules PRs to be merged to work correctly - but if you want to test this locally in the mean time you can replace that one URL with this code:

```php
switch ($version) {
    case '5':
        $modulesUrl = 'https://raw.githubusercontent.com/silverstripe/supported-modules/e9c9d44a6833bda035d6a3182effd167ab499549/modules.json';
        break;
    case '4':
        $modulesUrl = 'https://raw.githubusercontent.com/silverstripe/supported-modules/daa19e8c279f599088ba12b2a91b9a51cfbc5eb2/modules.json';
        break;
    case '3':
        $modulesUrl = 'https://raw.githubusercontent.com/silverstripe/supported-modules/30ceab854f964f4488f2834b80fc4d56f464e9c2/modules.json';
        break;
}
```

File: src/Console/CheckoutCommand.php [hunk 2]
```diff
@@ -11,50 +14,203 @@
...
+    /**
+     * Regex patterns for github repositories that should be ignored.
+     * These are repositories that don't have relevant API.
+     */
+    private array $ignoreModulePatterns = [
+        '/^silverstripe\/developer-docs$/',
+        '/^silverstripe\/installer$/',
+        '/^silverstripe\/vendor-plugin$/',
+        '/^silverstripe\/recipe-.*/',
+        '/-theme$/',
```
Comment (GuySartorelli) [inline src/Console/CheckoutCommand.php:26]: These are repositories which either have no PHP code at all, or their PHP code isn't really relevant.
The second part of that might be fairly subjective though.... feel free to recommend changes here if you think of any changes that are appropriate.

File: makedoc.sh [hunk 3]
```diff
@@ -4,7 +4,6 @@
 cd $(dirname "$0");
 set -e
 
-bin/docs checkout -v
-
-DOCTUM_COMPOSER_AUTOLOAD_FILE='./vendor/autoload.php' ./vendor/bin/doctum.php --version
-DOCTUM_COMPOSER_AUTOLOAD_FILE='./vendor/autoload.php' ./vendor/bin/doctum.php --ignore-parse-errors --force -v update conf/doctum.php
+bin/docs checkout -v \
+ && DOCTUM_COMPOSER_AUTOLOAD_FILE='./vendor/autoload.php' ./vendor/bin/doctum.php --version \
+ && DOCTUM_COMPOSER_AUTOLOAD_FILE='./vendor/autoload.php' ./vendor/bin/doctum.php --ignore-parse-errors --force -v update conf/doctum.php
```
Comment (GuySartorelli) [inline makedoc.sh:9]: We shouldn't try the doctum code if we failed to build out the version map.

File: composer.json [hunk 4]
```diff
@@ -24,7 +24,7 @@
         }
     },
     "require-dev": {
-        "phpunit/phpunit": "^7",
+        "phpunit/phpunit": "^8.5",
```
Comment (GuySartorelli) [inline composer.json:27]: Needed to use `$this->expectWarning()` in a test.

File: tests/SilverStripeRemoteRepositoryTest.php [hunk 5]
```diff
@@ -3,75 +3,161 @@
...
+            file_put_contents($filePath, self::$fileContent);
+        }
+    }
 
-        $url = $remoteRepo->getFileUrl(
-            '3',
-            'silverstripe/graphql/src/Extensions/IntrospectionProvider.php',
-            0
-        );
+    public function provideGetRepoUrl()
```
Comment (GuySartorelli) [inline tests/SilverStripeRemoteRepositoryTest.php:91]: These are all the same tests that were happening before (except the graphql 3 one which was silly - that's not a valid mapping) plus one for CMS 5 and 3 just so we have one for each of the three mapped versions.

File: tests/SilverStripeRemoteRepositoryTest.php [hunk 6]
```diff
@@ -3,75 +3,161 @@
...
-        );
+    public static function setUpBeforeClass(): void
+    {
+        $config = Config::getConfig();
+        $filePath = Config::configPath($config['paths']['versionmap']);
+        if (file_exists($filePath)) {
+            self::$fileContent = file_get_contents($filePath);
+            unlink($filePath);
+        }
+        file_put_contents($filePath, json_encode([
```
Comment (GuySartorelli) [inline tests/SilverStripeRemoteRepositoryTest.php:21]: We need to make sure there's some mapping available to test against. But if you're running tests locally you may already have a map there so we make sure to keep that and restore it after the test.

File: src/Console/CheckoutCommand.php [hunk 7]
```diff
@@ -11,50 +14,204 @@
...
+     * These are repositories that don't have relevant API.
+     */
+    private array $ignoreModulePatterns = [
+        '/^silverstripe\/developer-docs$/',
+        '/^silverstripe\/installer$/',
+        '/^silverstripe-themes\/simple$/',
+        '/^silverstripe\/vendor-plugin$/',
+        '/^silverstripe\/recipe-.*/',
+        '/-theme$/',
+    ];
```
Comment (GuySartorelli) [inline src/Console/CheckoutCommand.php:28]: These are repositories which either have no PHP code at all, or their PHP code isn't really relevant.
The second part of that might be fairly subjective though.... feel free to recommend changes here if you think of any changes that are appropriate.

**Review by emteknetnz — APPROVED** [review]

**Pull request merged**